README
App-PasswdMgr

The README is used to introduce the module and provide instructions on
how to install the module, any machine dependencies it may have (for
example C compilers and installed libraries) and any other information
that should be provided before the module is installed.

A README file is required for CPAN modules since CPAN extracts the README
file from a module distribution so that people browsing the archive
can use it to get an idea of the module's uses. It is usually a good idea
to provide version information here so that people can decide whether
fixes for the module are worth downloading.

INSTALLATION

To install this module, run the following commands:

    perl Build.PL
    ./Build
    ./Build test
    ./Build install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc App::PasswdMgr

You can also look for information at:

    RT, CPAN's request tracker
        http://rt.cpan.org/NoAuth/Bugs.html?Dist=App-PasswdMgr

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/App-PasswdMgr

    CPAN Ratings
        http://cpanratings.perl.org/d/App-PasswdMgr

    Search CPAN
        http://search.cpan.org/dist/App-PasswdMgr/

    Source Code
        git://github.com/ivanwills/App-PasswdMgr.git

COPYRIGHT AND LICENCE

Copyright (C) 2016 Ivan Wills

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for App-PasswdMgr

0.0.1   2016-06-21
        First version, released on an unsuspecting world.



lib/App/PasswdMgr.pm
package App::PasswdMgr;

# Created on: 2016-06-17 06:58:25
# Create by:  Ivan Wills
# $Id$
# $Revision$, $HeadURL$, $Date$
# $Revision$, $Source$, $Date$

use Moo;
use warnings;
use version;
use Carp;
use English qw/ -no_match_vars /;
use Getopt::Alt;
use Path::Tiny;
use YAML::Syck qw/ Load Dump /;
use Crypt::CBC;
use Crypt::Cipher::Blowfish;
use Digest::SHA qw/sha512_base64/;
use IO::Prompt;
use App::PasswdMgr::List;

our $VERSION     = version->new('0.0.1');

has [qw/cbc opt data/] => ( is => 'rw' );

sub run {
    my ($self) = @_;

    my ($options, $cmd, $opt) = get_options(
        {
            name        => 'passwdmgr',
            conf_prefix => '.',
            helper      => 1,
            default     => {
                test   => 0,
                passwords => "$ENV{HOME}/.passwdmgr",
            },
        },
        [
            'passwords|p=s',
            'dump|d',
            'create',
            'force',
        ],
    );

    $self->opt($options);
    if ( $options->create || !-f $self->opt->passwords ) {
        $self->create;
    }
    else {
        $self->build_cbc();
    }

    $self->data( Load( $self->cbc->decrypt( scalar path($self->opt->passwords)->slurp )));

    if ( !ref $self->data || ! $self->data->can('show') ) {
        warn "Bad password!\n";
        return;
    }
    if ( $options->dump ) {
        print Dump($self->data);
        exit 1;
    }

    $self->data->show;

    path($self->opt->passwords)->spew( scalar $self->cbc->encrypt( Dump($self->data) ) );

    return;
}

sub build_cbc {
    my ($self, $confirm) = @_;

    my $key = prompt( -p => 'Password: ', -e => '*' ) || die "No key!";
    my $iv  = Crypt::CBC->random_bytes(8);

    if ($confirm) {
        my $confirm = prompt( -p => 'Re-enter password: ', -e => '*' );

        die "Passwords don't match!\n" if $key ne $confirm;
    }

    $self->cbc(
        Crypt::CBC->new(
            -cipher => 'Cipher::Blowfish',
            -key    => $key,
        )
    );

    return;
}

sub create {
    my ($self) = @_;

    if ( -f $self->opt->passwords && $self->opt->force ) {
        die "Won't try to replace existing file '" . $self->opt->passwords ."'!\n";
    }

    $self->build_cbc(1);

    path($self->opt->passwords)->spew(
        scalar $self->cbc->encrypt(
            Dump(App::PasswdMgr::List->new(contents => {}, name => 'Base Menu' ))
        )
    );
}


1;

__END__

=head1 NAME

App::PasswdMgr - A command line password manager

=head1 VERSION

This documentation refers to App::PasswdMgr version 0.0.1

=head1 SYNOPSIS

    passwdmgr [options]
    passwdmgr (-p|--passwords) password-file

  OPTIONS:
    -p --passwords[=]file
                    File with stored passwords (Default ~/.passwdmgr)
    -c --create     Create new passwords file
    -f --force      Force the overwritting of existing passwords file
    -v --verbose    Show verbose information
       --version    Show passwdmrg version
       --help       Show this help
       --man        Show full help documentation

=head1 DESCRIPTION

=head1 SUBROUTINES/METHODS

=head2 C<run ()>

Runs the whole thing

=head2 C<build_cbc ( $confirm )>

Set up the decryption object

=head2 C<create ()>

Create a new passwordmgr file.

=head1 ATTRIBUTES

=over 4

=item cbc

The L<Crypt::CBC> object for encrypting the data.

=item opt

Command line optons.

=item data

The internal decrypted passwords.

=back

=head1 DIAGNOSTICS

=head1 CONFIGURATION AND ENVIRONMENT

=head1 DEPENDENCIES

=head1 INCOMPATIBILITIES

=head1 BUGS AND LIMITATIONS

There are no known bugs in this module.

Please report problems to Ivan Wills (ivan.wills@gmail.com).

Patches are welcome.

=head1 AUTHOR

Ivan Wills - (ivan.wills@gmail.com)

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2016 Ivan Wills (14 Mullion Close, Hornsby Heights, NSW Australia 2077).
All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See L<perlartistic>.  This program is
distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

=cut


lib/App/PasswdMgr/Base.pm
package App::PasswdMgr::Base;

# Created on: 2016-06-20 18:14:54
# Create by:  Ivan Wills
# $Id$
# $Revision$, $HeadURL$, $Date$
# $Revision$, $Source$, $Date$

use Moo;
use warnings;
use version;
use Carp;
use Scalar::Util;
use List::Util;
#use List::MoreUtils;
use Data::Dumper qw/Dumper/;
use English qw/ -no_match_vars /;
use IO::Prompt;
use Term::Size::Any qw/chars/;
use Term::Screen;

our $VERSION = version->new('0.0.1');

has contents => (
    is      => 'rw',
    default => sub {{}},
);
has parent => ( is => 'rw' );
has name   => ( is => 'rw' );

sub show {
    my ($self, $retry) = @_;

    if ( ! $retry ) {
        $self->clear;
    }

    my %actions = (
        q => {
            description => "Quit",
        },
        %{ $self->actions },
    );

    my $i = 0;
    for my $content (sort {lc $a cmp lc $b} keys %{ $self->contents }) {
        my $suffix = $self->suffix($content);
        $actions{$i++} = {
            description => "$content$suffix",
            content     => $content,
        };
    }

    my $menu = $self->menu(%actions);
    print $menu;

    my $ans = $self->question(
        -p => '> ',
        ( $i <= 10 ? '-one_char' : () ),
    );
    $self->clear;

    return if $ans eq '' || $ans eq 'q';

    my $method = exists $actions{$ans} && $actions{$ans}{method};

    return $self->$method() if $method;

    if ( ! exists $actions{$ans} ) {
        my @actions = sort keys %actions;
        print "\nPlease choose one of " . ( join ', ', @actions[0..$#actions-1] ) . " or $actions[-1]!\n";
        $retry++;
        return if $retry > 10;
        return $self->show($retry);
    }

    $self->contents->{$actions{$ans}{content}}->can('show')
        ? $self->contents->{$actions{$ans}{content}}->show
        : $self->edit($actions{$ans}{content});

    return $self->show;
}

sub clear {
    my ($cols, $rows) = chars();
    Term::Screen->new->at(0,0)->puts(' ' x ( $rows * $cols ));
}

sub menu {
    my ($self, %menu) = @_;
    my @ordered = sort {$a =~ /^\d+$/ && $b =~ /^\d+$/ ? $a <=> $b : $a cmp $b} keys %menu;

    print $self->name && $self->full_name || "Select one :", "\n";

    my $menu = '';
    for my $item (@ordered) {
        $menu .= sprintf "\t%2s  %s\n", $item, $menu{$item}{description};
    }

    return $menu;
}

sub full_name {
    my ($self) = @_;

    return $self->name if ! $self->parent;

    return $self->parent->full_name . ' / ' . $self->name . $self->parent->suffix($self->name);
}

sub suffix {''}

sub rename {
    my ($self) = @_;

    print $self->full_name . "\n";
    my $new_name = $self->question('New name: ');
    if ( $new_name ) {
        if ( $self->parent->contents->{$new_name} ) {
            print "A '$new_name' already exists!\n";
        }
        else {
            my $old_name = $self->name;
            $self->name($new_name);
            $self->parent->contents->{$new_name} = $self;
            delete $self->parent->contents->{$old_name};
        }
    }

    return $self->show;
}

sub _new {
    my ($self, $text, $type) = @_;
    my $name = $self->question( -p => $text );

    if ( ! $name ) {
        return $self->show(1);
    }
    elsif ( $name =~ /^_/ ) {
        print "Can't start with underscores!\n";
        return $self->show(1);
    }
    elsif ( exists $self->contents->{$name} ) {
        $self->clear;
        print "'$name' already exists!\n";
        return $self->show(1);
    }

    $self->contents->{$name} = $type->new(
        parent => $self,
        name   => $name,
    );

    $self->contents->{$name}->show;

    return $self->show;
}

sub question {
    my ($self, @args) = @_;
    return '' . prompt(@args);
}

1;

__END__

=head1 NAME

App::PasswdMgr::Base - Base class for object stored by App::PasswdMgr

=head1 VERSION

This documentation refers to App::PasswdMgr::Base version 0.0.1


=head1 SYNOPSIS

   use App::PasswdMgr::Base;

   # Brief but working code example(s) here showing the most common usage(s)
   # This section will be as far as many users bother reading, so make it as
   # educational and exemplary as possible.

=head1 DESCRIPTION

=head1 SUBROUTINES/METHODS

=head2 C<show ()>

=head2 C<clear ()>

=head2 C<menu ()>

=head2 C<full_name ()>

=head2 C<suffix ()>

=head2 C<rename ()>

=head2 C<_new ()>

=head2 C<question ()>

=head1 ATTRIBUTES

=over 4

=item contents

Stores the item's data

=item parent

A reference to the parent item

=item name

The name of the item

=back

=head1 DIAGNOSTICS

=head1 CONFIGURATION AND ENVIRONMENT

=head1 DEPENDENCIES

=head1 INCOMPATIBILITIES

=head1 BUGS AND LIMITATIONS

There are no known bugs in this module.

Please report problems to Ivan Wills (ivan.wills@gmail.com).

Patches are welcome.

=head1 AUTHOR

Ivan Wills - (ivan.wills@gmail.com)

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2016 Ivan Wills (14 Mullion Close, Hornsby Heights, NSW Australia 2077).
All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See L<perlartistic>.  This program is
distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

=cut


lib/App/PasswdMgr/List.pm
package App::PasswdMgr::List;

# Created on: 2016-06-20 17:29:11
# Create by:  Ivan Wills
# $Id$
# $Revision$, $HeadURL$, $Date$
# $Revision$, $Source$, $Date$

use Moo;
use warnings;
use version;
use Carp;
use Scalar::Util;
use List::Util;
#use List::MoreUtils;
use Data::Dumper qw/Dumper/;
use English qw/ -no_match_vars /;
use IO::Prompt;
use App::PasswdMgr::Password;

extends 'App::PasswdMgr::Base';

our $VERSION = version->new('0.0.1');

sub actions {
    my ($self) = @_;
    my $actions = {
        p => {
            description => 'New password',
            method      => 'new_password',
        },
        g => {
            description => 'New group',
            method      => 'new_group',
        },
    };
    if ( $self->parent ) {
        $actions->{d} = {
            description => 'Delete this group (and ' . (keys %{$self->contents}) . ' items)',
            method      => 'delete',
        };
        $actions->{r} = {
            description => 'Rename group',
            method      => 'rename',
        };
    }
    return $actions;
}

sub suffix {
    my ($self, $content) = @_;
    return $content eq 'name'                            ? undef
        : ref $self->contents->{$content} eq __PACKAGE__ ? ''
        :                                                  ' (password)';
}

sub new_group {
    my ($self) = @_;
    return $self->_new( "New groups name: ", 'App::PasswdMgr::List' );
}

sub new_password {
    my ($self) = @_;
    return $self->_new( "New password's name: ", 'App::PasswdMgr::Password' );
}

sub delete {
    my ($self) = @_;

    if ( keys %{ $self->contents } ) {
        my $delete = prompt( -p => 'Really delete? [yn] ', '-yes' );
        return $self->show if ! $delete;
    }

    delete $self->parent->contents->{$self->name};

    return;
}

1;

__END__

=head1 NAME

App::PasswdMgr::List - Stores a list of PasswdMgr items

=head1 VERSION

This documentation refers to App::PasswdMgr::List version 0.0.1

=head1 SYNOPSIS

   use App::PasswdMgr::List;

   # Brief but working code example(s) here showing the most common usage(s)
   # This section will be as far as many users bother reading, so make it as
   # educational and exemplary as possible.


=head1 DESCRIPTION

A full description of the module and its features.

May include numerous subsections (i.e., =head2, =head3, etc.).


=head1 SUBROUTINES/METHODS

=head2 C<actions ()>

Returns a hash ref of available actions for the current list.

=head2 C<suffix ( $content )>

Returns the suffix '(password)' if the item is a L<App::PasswdMgr::Password>
item.

=head2 C<new_group ()>

Creates new group L<App::PasswordMgr::List> of list items.

=head2 C<new_password ()>

Creates a new password item

=head2 C<delete ()>

Delets the current group from the parent list.

=head1 DIAGNOSTICS

A list of every error and warning message that the module can generate (even
the ones that will "never happen"), with a full explanation of each problem,
one or more likely causes, and any suggested remedies.

=head1 CONFIGURATION AND ENVIRONMENT

A full explanation of any configuration system(s) used by the module, including
the names and locations of any configuration files, and the meaning of any
environment variables or properties that can be set. These descriptions must
also include details of any configuration language used.

=head1 DEPENDENCIES

A list of all of the other modules that this module relies upon, including any
restrictions on versions, and an indication of whether these required modules
are part of the standard Perl distribution, part of the module's distribution,
or must be installed separately.

=head1 INCOMPATIBILITIES

A list of any modules that this module cannot be used in conjunction with.
This may be due to name conflicts in the interface, or competition for system
or program resources, or due to internal limitations of Perl (for example, many
modules that use source code filters are mutually incompatible).

=head1 BUGS AND LIMITATIONS

A list of known problems with the module, together with some indication of
whether they are likely to be fixed in an upcoming release.

Also, a list of restrictions on the features the module does provide: data types
that cannot be handled, performance issues and the circumstances in which they
may arise, practical limitations on the size of data sets, special cases that
are not (yet) handled, etc.

The initial template usually just has:

There are no known bugs in this module.

Please report problems to Ivan Wills (ivan.wills@gmail.com).

Patches are welcome.

=head1 AUTHOR

Ivan Wills - (ivan.wills@gmail.com)

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2016 Ivan Wills (14 Mullion Close, Hornsby Heights, NSW Australia 2077).
All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See L<perlartistic>.  This program is
distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

=cut


lib/App/PasswdMgr/Password.pm
package App::PasswdMgr::Password;

# Created on: 2016-06-20 17:29:26
# Create by:  Ivan Wills
# $Id$
# $Revision$, $HeadURL$, $Date$
# $Revision$, $Source$, $Date$

use Moo;
use warnings;
use version;
use Carp;
use Data::Dumper qw/Dumper/;
use English qw/ -no_match_vars /;
use App::PasswdMgr::Password::Param;
use App::PasswdMgr::Password::Encrypted;

extends 'App::PasswdMgr::Base';

our $VERSION = version->new('0.0.1');

sub actions {
    my ($self) = @_;
    my $actions = {
        n => {
            description => 'New parameter',
            method      => 'new_parameter',
        },
        p => {
            description => 'Enter password',
            method      => 'enter_password',
        },
        r => {
            description => 'Rename',
            method      => 'rename',
        },
    };

    if ( $self->contents->{password} ) {
        $actions = {
            %$actions,
            c => {
                description => 'Put password into clipboard',
                method      => 'clipboard',
            },
            s => {
                description => 'Show password',
                method      => 'show_password',
            },
        };
    }

    return $actions;
}

sub edit {
    my ($self, $content) = @_;

    return $self->show;
}

sub clipboard {
    my ($self, $content) = @_;

    $self->contents->{password}->clipboard(1);

    return $self->show;
}

before show => sub {
    my ($self) = @_;

    $self->new_parameter(1) if !%{ $self->contents };

    return;
};

sub new_parameter {
    my ($self, $hide) = @_;

    my $param = App::PasswdMgr::Password::Param->new(
        parent => $self,
    )->set;
    $self->contents->{$param->name} = $param;

    return $hide || $self->show;
}

sub enter_password {
    my ($self, $hide) = @_;

    if ( $self->contents->{password} ) {
        $self->contents->{password}->set;
    }
    else {
        $self->contents->{password} = App::PasswdMgr::Password::Param->new(
            parent => $self,
            name   => 'password',
        )->set;
    }

    return $hide || $self->show;
}

sub show_password {
    my ($self, $content) = @_;

    print $self->contents->{password}[-1]{text} . "\n";

    $self->question(-p => "Press the any key to continue", '-one_char');

    return $self->show;
}

1;

__END__

=head1 NAME

App::PasswdMgr::Password - <One-line description of module's purpose>

=head1 VERSION

This documentation refers to App::PasswdMgr::Password version 0.0.1


=head1 SYNOPSIS

   use App::PasswdMgr::Password;

   # Brief but working code example(s) here showing the most common usage(s)
   # This section will be as far as many users bother reading, so make it as
   # educational and exemplary as possible.


=head1 DESCRIPTION

A full description of the module and its features.

May include numerous subsections (i.e., =head2, =head3, etc.).


=head1 SUBROUTINES/METHODS

=head2 C<actions ()>

Returns a hash ref of available actions for the current password.

=head2 C<edit ()>

Change the name of the password

=head2 C<clipboard ()>

Copy the password to the clipboard.

=head2 C<before: show ()>

First time show is called this will get the user to fill out details.

=head2 C<new_parameter ()>

Create a new parameter (L<App::PasswdMgr::Password::Param>) for the password

=head2 C<enter_password ()>

Add/change the stored password

=head2 C<show_password ()>

Show the password unencrypted.

=head1 DIAGNOSTICS

A list of every error and warning message that the module can generate (even
the ones that will "never happen"), with a full explanation of each problem,
one or more likely causes, and any suggested remedies.

=head1 CONFIGURATION AND ENVIRONMENT

A full explanation of any configuration system(s) used by the module, including
the names and locations of any configuration files, and the meaning of any
environment variables or properties that can be set. These descriptions must
also include details of any configuration language used.

=head1 DEPENDENCIES

A list of all of the other modules that this module relies upon, including any
restrictions on versions, and an indication of whether these required modules
are part of the standard Perl distribution, part of the module's distribution,
or must be installed separately.

=head1 INCOMPATIBILITIES

A list of any modules that this module cannot be used in conjunction with.
This may be due to name conflicts in the interface, or competition for system
or program resources, or due to internal limitations of Perl (for example, many
modules that use source code filters are mutually incompatible).

=head1 BUGS AND LIMITATIONS

A list of known problems with the module, together with some indication of
whether they are likely to be fixed in an upcoming release.

Also, a list of restrictions on the features the module does provide: data types
that cannot be handled, performance issues and the circumstances in which they
may arise, practical limitations on the size of data sets, special cases that
are not (yet) handled, etc.

The initial template usually just has:

There are no known bugs in this module.

Please report problems to Ivan Wills (ivan.wills@gmail.com).

Patches are welcome.

=head1 AUTHOR

Ivan Wills - (ivan.wills@gmail.com)

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2016 Ivan Wills (14 Mullion Close, Hornsby Heights, NSW Australia 2077).
All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See L<perlartistic>.  This program is
distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

=cut


lib/App/PasswdMgr/Password/Encrypted.pm
package App::PasswdMgr::Password::Encrypted;

# Created on: 2016-06-23 06:55:30
# Create by:  Ivan Wills
# $Id$
# $Revision$, $HeadURL$, $Date$
# $Revision$, $Source$, $Date$

use Moo;
use warnings;
use version;
use Carp;
use Scalar::Util;
use List::Util;
#use List::MoreUtils;
use Data::Dumper qw/Dumper/;
use English qw/ -no_match_vars /;

extends 'App::PasswdMgr::Password::Param';

our $VERSION = version->new('0.0.1');

has '+types' => (
    default => sub {{
        'Auto generate'  => 'generate',
        'Enter password' => 'hand',
    }},
);
has '+value_question' => (
    default => sub {return {
        -p => 'Password: ',
        -e => '*',
    }},
);

sub actions {
    {
        c => {
            description => 'Put password into clipboard',
            method      => 'clipboard',
        },
        n => {
            description => 'Insert new parameter',
            method      => 'new_parameter',
        },
        p => {
            description => 'Enter password',
            method      => 'enter_password',
        },
        s => {
            description => 'Show password',
            method      => 'show_password',
        },
        r => {
            description => 'Rename group',
            method      => 'rename',
        },
    }
}

sub suffix {
    return '(********)';
}

sub enter_value {
    my ($self) = @_;
    return $self->type eq 'hand';
}

1;

__END__

=head1 NAME

App::PasswdMgr::Password::Encrypted - <One-line description of module's purpose>

=head1 VERSION

This documentation refers to App::PasswdMgr::Password::Encrypted version 0.0.1

=head1 SYNOPSIS

   use App::PasswdMgr::Password::Encrypted;

   # Brief but working code example(s) here showing the most common usage(s)
   # This section will be as far as many users bother reading, so make it as
   # educational and exemplary as possible.


=head1 DESCRIPTION

A full description of the module and its features.

May include numerous subsections (i.e., =head2, =head3, etc.).


=head1 SUBROUTINES/METHODS

=head2 C<actions ()>

Returns a hash ref of available actions

=head2 C<suffix ()>

Adds a '(*****)' to passwords

=head2 C<enter_value ()>

Returns true if the password is to be a user entered value

=head1 ATTRIBUTES

=over 4

=item types

Returns the menu of password types

=item value_question

Returns the parameters for asking to enter a password

=back

=head1 DIAGNOSTICS

A list of every error and warning message that the module can generate (even
the ones that will "never happen"), with a full explanation of each problem,
one or more likely causes, and any suggested remedies.

=head1 CONFIGURATION AND ENVIRONMENT

A full explanation of any configuration system(s) used by the module, including
the names and locations of any configuration files, and the meaning of any
environment variables or properties that can be set. These descriptions must
also include details of any configuration language used.

=head1 DEPENDENCIES

A list of all of the other modules that this module relies upon, including any
restrictions on versions, and an indication of whether these required modules
are part of the standard Perl distribution, part of the module's distribution,
or must be installed separately.

=head1 INCOMPATIBILITIES

A list of any modules that this module cannot be used in conjunction with.
This may be due to name conflicts in the interface, or competition for system
or program resources, or due to internal limitations of Perl (for example, many
modules that use source code filters are mutually incompatible).

=head1 BUGS AND LIMITATIONS

A list of known problems with the module, together with some indication of
whether they are likely to be fixed in an upcoming release.

Also, a list of restrictions on the features the module does provide: data types
that cannot be handled, performance issues and the circumstances in which they
may arise, practical limitations on the size of data sets, special cases that
are not (yet) handled, etc.

The initial template usually just has:

There are no known bugs in this module.

Please report problems to Ivan Wills (ivan.wills@gmail.com).

Patches are welcome.

=head1 AUTHOR

Ivan Wills - (ivan.wills@gmail.com)

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2016 Ivan Wills (14 Mullion Close, Hornsby Heights, NSW Australia 2077).
All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See L<perlartistic>.  This program is
distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

=cut


lib/App/PasswdMgr/Password/Param.pm
package App::PasswdMgr::Password::Param;

# Created on: 2016-06-22 07:19:55
# Create by:  Ivan Wills
# $Id$
# $Revision$, $HeadURL$, $Date$
# $Revision$, $Source$, $Date$

use Moo;
use warnings;
use version;
use Carp;
use English qw/ -no_match_vars /;
use Clipboard qw//;

extends 'App::PasswdMgr::Base';

our $VERSION = version->new('0.0.1');

has [qw/value type/] => ( is => 'rw' );
has types => (
    is      => 'rw',
    default => sub {{
        Username => 'username',
        Key      => 'key',
        Other    => 'other',
    }},
);
has value_question => (
    is      => 'rw',
    default => sub {return {
        -p => 'Value: ',
    }},
);

sub actions {
    {
        c => {
            description => 'Copy into clipboard',
            method      => 'clipboard',
        },
        n => {
            description => 'Insert new parameter',
            method      => 'new_parameter',
        },
        p => {
            description => 'Enter password',
            method      => 'enter_password',
        },
        s => {
            description => 'Show password',
            method      => 'show_password',
        },
        r => {
            description => 'Rename group',
            method      => 'rename',
        },
    }
}

sub set {
    my ($self) = @_;

    my$name;
    my $type = $self->question(
        -p => 'Parameter type: ',
        -m => $self->types,
    );

    if ( $type eq 'other' ) {
        $name = $self->question('Name: ');
    }
    else {
        $name = ucfirst $type;
    }
    $self->name($name) if !$self->name;
    $self->type($type);

    if ($self->enter_value) {
        my $value = $self->question(%{ $self->value_question });
        $self->value($value);
    }

    return $self;
}

sub enter_value {1}

sub suffix {
    my ($self) = @_;

    return '(' . $self->value . ')';
}

sub clipboard {
    my ($self, $hide) = @_;

    Clipboard->copy( $self->value );

    return $hide || $self->show;
}

1;

__END__

=head1 NAME

App::PasswdMgr::Password::Param - <One-line description of module's purpose>

=head1 VERSION

This documentation refers to App::PasswdMgr::Password::Param version 0.0.1

=head1 SYNOPSIS

   use App::PasswdMgr::Password::Param;

   # Brief but working code example(s) here showing the most common usage(s)
   # This section will be as far as many users bother reading, so make it as
   # educational and exemplary as possible.


=head1 DESCRIPTION

A full description of the module and its features.

May include numerous subsections (i.e., =head2, =head3, etc.).


=head1 SUBROUTINES/METHODS

=head2 C<actions ()>

Returns a hash ref of available actions

=head2 C<suffix ()>

Adds the value when displaying the parameter.

=head2 C<enter_value ()>

Returns true if the password is to be a user entered value

=head2 C<clipboard ()>

Copy's the parameters value to the system clipboard.

=head2 C<set ()>

This method is used to allow the user to set/update the parameter

=head1 ATTRIBUTES

=over 4

=item types

Returns the menu of password types

=item type

Stores the type of parameter

=item value

Stores the value of the parameter

=item value_question

Returns the parameters for asking to enter a password

=back

=head1 DIAGNOSTICS

A list of every error and warning message that the module can generate (even
the ones that will "never happen"), with a full explanation of each problem,
one or more likely causes, and any suggested remedies.

=head1 CONFIGURATION AND ENVIRONMENT

A full explanation of any configuration system(s) used by the module, including
the names and locations of any configuration files, and the meaning of any
environment variables or properties that can be set. These descriptions must
also include details of any configuration language used.

=head1 DEPENDENCIES

A list of all of the other modules that this module relies upon, including any
restrictions on versions, and an indication of whether these required modules
are part of the standard Perl distribution, part of the module's distribution,
or must be installed separately.

=head1 INCOMPATIBILITIES

A list of any modules that this module cannot be used in conjunction with.
This may be due to name conflicts in the interface, or competition for system
or program resources, or due to internal limitations of Perl (for example, many
modules that use source code filters are mutually incompatible).

=head1 BUGS AND LIMITATIONS

A list of known problems with the module, together with some indication of
whether they are likely to be fixed in an upcoming release.

Also, a list of restrictions on the features the module does provide: data types
that cannot be handled, performance issues and the circumstances in which they
may arise, practical limitations on the size of data sets, special cases that
are not (yet) handled, etc.

The initial template usually just has:

There are no known bugs in this module.

Please report problems to Ivan Wills (ivan.wills@gmail.com).

Patches are welcome.

=head1 AUTHOR

Ivan Wills - (ivan.wills@gmail.com)

=head1 LICENSE AND COPYRIGHT

Copyright (c) 2016 Ivan Wills (14 Mullion Close, Hornsby Heights, NSW Australia 2077).
All rights reserved.

This module is free software; you can redistribute it and/or modify it under
the same terms as Perl itself. See L<perlartistic>.  This program is
distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.

=cut
